### ISSUE TITLE
isinstance(RatingScale(win), RatingScale) is False

### ISSUE THREAD

**Issue by Kodiologist** [psychopy/psychopy#2064]
On PsychoPy 30066a534d532e070518c7abf93f3175f4bd04ef and Python 2.7.15rc1:

    from psychopy.visual import Window, RatingScale
    win = Window()
    print isinstance(RatingScale(win), RatingScale)  # False

This seems to be due to the use of lazy imports in `psychopy.visual`. If you import `RatingScale` from `psychopy.visual.ratingscale` instead of `psychopy.visual`, or write `from psychopy import visual` and refer to the class as `visual.RatingScale`, the `isinstance` comes out true, as expected.

**Comment by peircej** [comment]
Yes, I think that's due to the lazy importing. I'm not sure that much can be done about it. When you do `from psychopy.visual import Window, RatingScale` you don't get a RatingScale class but a wrapper pointing to it, so Python is correct. I know it's annoying that this gives a surprising result, but the alternatives would be more painful: if we don't do the lazy import then either:

- all stimuli have to be imported from their full paths like `from psychopy.visual.ratingscale import RatingScale` which takes a line of importing for every stimulus
- all stimuli would be imported (not lazily) by visual and that means that `import visual` is slow

I guess we could leave this open for a week or two to see if other ideas come up, but I think it should probably be closed as unfixable

**Comment by Kodiologist** [comment]
Shoot. You could put a warning in the documentation about this, but a lot of people would simply miss it, of course.

It might make sense for `psychopy.contrib.lazy_import` to provide its own version of `isinstance` that works the same as the base function for non-lazy classes, but does whatever is necessary to get the real underlying class for lazy classes.

**Comment by mdcutone** [comment]
@peircej 

I'm not sure the classes included with PsychoPy are 'heavy' enough to justify lazy loading (at least on modern hardware). It might be something to look into.

**Comment by peircej** [comment]
The *original* reason for the lazy loading wasn't really for performance. The issue was that some stimuli have dependencies that don't play nice (like avbin in MovieStim, vlc in MovieStim2, TextBox had some issue that I've forgotten). Users would get an error message about avbin failing even though they didn't need it in their experiment (no movies playing) and this would cause them to waste hours trying to solve an irrelevant problem.

There is also the question of performance in that, although each class is small, we have a lot of classes, constantly growing (Button, TextBox2, Form all recently added/coming). PsychoPy experiments already take quite a long time to launch on their first run (less 2nd time due to caching), so I feel like we should try a little and reduce that problem.

If I could go back I would change the imports to use the long explicit import:
```
from psychopy.visual.ratingscale import RatingScale
```
so people can be selective and have no lazy-loading, but most people haven't been using that so lots of code will break in the switch.

**Comment by Kodiologist** [comment]
> The *original* reason for the lazy loading wasn't really for performance. The issue was that some stimuli have dependencies that don't play nice (like avbin in MovieStim, vlc in MovieStim2, TextBox had some issue that I've forgotten). Users would get an error message about avbin failing even though they didn't need it in their experiment (no movies playing) and this would cause them to waste hours trying to solve an irrelevant problem.

There's an easy solution to that without formal lazy loading, which is to put the `import`s for the dependencies in the constructors and assign to a global, instead of using `import` at the top level. Something like:

```python
class MyStim:
   def __init__(self):
       global some_dependency
       import some_dependency
       …
   …
```

Perhaps it's worth trying that and removing lazy loading and seeing if the startup time for simple experiments is still acceptable.

**Comment by mdcutone** [comment]
> The _original_ reason for the lazy loading wasn't really for performance. The issue was that some stimuli have dependencies that don't play nice (like avbin in MovieStim, vlc in MovieStim2, TextBox had some issue that I've forgotten). Users would get an error message about avbin failing even though they didn't need it in their experiment (no movies playing) and this would cause them to waste hours trying to solve an irrelevant problem.

That's a completely valid reason for lazy loading.

> so people can be selective and have no lazy-loading, but most people haven't been using that so lots of code will break in the switch.

I agree, let's not break anything. 

@Kodiologist 
>There's an easy solution to that without formal lazy loading, which is to put the imports for the dependencies in the constructors and assign to a global, instead of using import at the top level. Something like:
> Perhaps it's worth trying that and removing lazy loading and seeing if the startup time for simple experiments is still acceptable.

Good suggestion, I think that's the recommended way to do so. See "Import Statement Overhead" -- https://wiki.python.org/moin/PythonSpeed/PerformanceTips

**Comment by peircej** [comment]
That's a nice idea. I'm up for testing it, at least.

**Comment by mdcutone** [comment]
This seems to work. Reminds me of the `requires` function in R so I named it that...

```
import importlib
import sys

class Test(object):
    def __init__(self, *args, **kwargs):
        requires('math')

def requires(mod_name):
    if mod_name not in sys.modules:
        globals()[mod_name] = importlib.import_module(mod_name)
    
def main(args):
    m = Test()
    print(math.cos(10))
    return 0
    
if __name__ == '__main__':
    import sys
    sys.exit(main(sys.argv))

```
